lib/Conversion/ExportVerilog/PrepareForEmission.cpp

```diff
@@ -566,7 +566,9 @@ void ExportVerilog::prepareHWModule(Block &block,
       // it, we are good to generate a wire.
       if (!isProceduralRegion ||
           (isProceduralRegion && hoistNonSideEffectExpr(&op))) {
-        lowerUsersToTemporaryWire(op);
+        // If op is moved to a non-procedural region, create a temporary wire.
+        if (!op.getParentOp()->hasTrait<ProceduralRegion>())
```

Comment:
I had originally thought hoistNonSideEffectExpr should find a non-procedural parent or return false: https://github.com/llvm/circt/blob/041a5dc6a4f1e43eb1e50d90230d354414473b6a/lib/Conversion/ExportVerilog/PrepareForEmission.cpp#L374. Is this second check necessary for some other case?

Reply:
Yes, it is necessary because `hoistNonSideEffectExpr` also returns true when it moves the operation to a procedural region. 
https://github.com/llvm/circt/blob/041a5dc6a4f1e43eb1e50d90230d354414473b6a/lib/Conversion/ExportVerilog/PrepareForEmission.cpp#L402-L405

Reply:
I guess `hoistNonSideEffectExpr` is hoisting only one-level because we cannot always hoist operations to the toplevel: we have to resolve dependencies as well so we are processing regions in the post order. For example, in the example below we visit `%b` first but `%b` cannot be directly hoisted to toplevel because `%a` is not hoisted yet.  
```
sv.initial {
  sv.ifdef.procedural "A" {
    %a = ...
    sv.ifdef.procedural "B" {
       %b = comb.add %a 
    }
  } 
}
```